File: kf_gins_ros2/kf_gins_ros2/geo.py
import math
WGS84_A = 6378137.0
WGS84_E2 = 6.69437999014e-3
def llh_to_ecef(lat_rad, lon_rad, h_m):
    a=WGS84_A; e2=WGS84_E2
    sl=math.sin(lat_rad); cl=math.cos(lat_rad)
    slo=math.sin(lon_rad); clo=math.cos(lon_rad)
    N=a/(1.0 - e2*sl*sl)**0.5
    x=(N+h_m)*cl*clo; y=(N+h_m)*cl*slo; z=(N*(1.0-e2)+h_m)*sl
    return x,y,z
def ecef_to_enu(x,y,z,lat0_rad,lon0_rad,x0,y0,z0):
    sl=math.sin(lat0_rad); cl=math.cos(lat0_rad)
    slo=math.sin(lon0_rad); clo=math.cos(lon0_rad)
    dx=x-x0; dy=y-y0; dz=z-z0
    e=-slo*dx+clo*dy
    n=-clo*sl*dx - slo*sl*dy + cl*dz
    u= clo*cl*dx + slo*cl*dy + sl*dz
    return e,n,u
def rpy_to_quat(roll,pitch,yaw):
    cr=math.cos(roll*0.5); sr=math.sin(roll*0.5)
    cp=math.cos(pitch*0.5); sp=math.sin(pitch*0.5)
    cy=math.cos(yaw*0.5); sy=math.sin(yaw*0.5)
    w=cr*cp*cy + sr*sp*sy
    x=sr*cp*cy - cr*sp*sy
    y=cr*sp*cy + sr*cp*sy
    z=cr*cp*sy - sr*sp*cy
    return x,y,z,w

File: kf_gins_ros2_native/src/geo.cpp
#include "kf_gins_ros2_native/geo.hpp"
#include <cmath>
namespace geo {
void llh_to_ecef(double lat_rad, double lon_rad, double h, double& x, double& y, double& z){
  const double a=WGS84_A, e2=WGS84_E2;
  const double sl=std::sin(lat_rad), cl=std::cos(lat_rad);
  const double slo=std::sin(lon_rad), clo=std::cos(lon_rad);
  const double N=a/std::sqrt(1.0 - e2*sl*sl);
  x=(N+h)*cl*clo; y=(N+h)*cl*slo; z=(N*(1.0-e2)+h)*sl;
}
Eigen::Vector3d ecef_to_enu(const Eigen::Vector3d& ecef,const Eigen::Vector3d& ecef0,double lat0,double lon0){
  const double sl=std::sin(lat0), cl=std::cos(lat0);
  const double slo=std::sin(lon0), clo=std::cos(lon0);
  Eigen::Vector3d d = ecef - ecef0;
  return {-slo*d.x()+clo*d.y(), -clo*sl*d.x()-slo*sl*d.y()+cl*d.z(), clo*cl*d.x()+slo*cl*d.y()+sl*d.z()};
}
Eigen::Quaterniond rpy_to_quat(double r,double p,double y){
  const double cr=std::cos(r*0.5), sr=std::sin(r*0.5);
  const double cp=std::cos(p*0.5), sp=std::sin(p*0.5);
  const double cy=std::cos(y*0.5), sy=std::sin(y*0.5);
  Eigen::Quaterniond q; q.w()=cr*cp*cy+sr*sp*sy; q.x()=sr*cp*cy-cr*sp*sy; q.y()=cr*sp*cy+sr*cp*sy; q.z()=cr*cp*sy-sr*cp*cy; return q;
}
}
